sources/api/apiserver/src/server/controller.rs

```diff
@@ -420,21 +498,91 @@ pub(crate) fn get_metadata_for_data_keys<D: DataStore, S: AsRef<str>>(
     Ok(result)
 }
 
-/// Gets the value of a metadata key everywhere it's found in the data store.  Returns a mapping
-/// of data key to the metadata value associated with the requested key.
-pub(crate) fn get_metadata_for_all_data_keys<D: DataStore, S: AsRef<str>>(
+// Sort the given Hashmap in topological order
+fn get_sorted_metadata(
+    metadata: HashMap<Key, HashMap<Key, String>>,
+) -> Vec<(Key, HashMap<Key, String>)> {
+    let mut metadata_sorted: Vec<_> = metadata.into_iter().collect();
+    metadata_sorted.sort_by(|(k1, _), (k2, _)| {
+        k1.segments()
+            .len()
+            .cmp(&k2.segments().len())
+            .then_with(|| k1.name().cmp(k2.name()))
+    });
+    metadata_sorted.to_vec()
+}
+
+// Find the successor of the given setting and add the metadata
+// for the child at given depth in setting generator
+fn get_setting_generator_for_child<D: DataStore>(
+    datastore: &D,
+    data_key: &Key,
+    raw_setting_generator: RawSettingsGenerator,
+    result: &mut HashMap<String, Value>,
+) -> Result<()> {
+    // Get the leaf of the data key
+    let segments = data_key.segments();
+    let data_key_for_metadata = segments.last().context(error::InvalidKeyPairSnafu {
+        input: data_key.name(),
+    })?;
+
+    // Extract the prefix leaving the leaf
+    let new_data_key =
+        Key::from_segments(KeyType::Data, &segments[..segments.len().saturating_sub(1)]).context(
+            error::DataStoreSnafu {
+                op: format!("Unable to create key from segments: {:?}", segments),
+            },
+        )?;
+
+    let populated_segments_for_prefix = datastore
+        .list_populated_keys(new_data_key.name(), &Committed::Live)
+        .context(error::DataStoreSnafu {
+            op: format!(
+                "Unable to get the populated keys for given prefix: {}",
+                new_data_key.name()
+            ),
+        })?;
+
+    let depth = raw_setting_generator.depth;
+    let metadata_value = serde_json::to_value(SettingsGenerator::from(raw_setting_generator))
+        .context(error::SerializeSnafu)?;
+
+    for populated_segment in populated_segments_for_prefix {
+        // Extract the successor of the prefix
+        let mut new_segments = populated_segment
```

Comment:
Do we need to check that the length of the populated segment is at least the length of the prefix plus the depth?

Otherwise, if we have a field on the parent that's populated, it might be found here.

```
settings.host-containers.blah = "..."
settings.host-containers.admin.enabled = true
```

If we had a source generator for `settings.host-containers` with a depth of 1, I think this would attempt to generate:
```
settings.host-containers.source
```

Reply:
This would be a good case to add to the unit tests.